_posts/2017-07-05-SpringIO17-Summary.md

```diff
@@ -0,0 +1,773 @@
+As always, Michael delivered: he made me realize **there are still plenty of PL/SQL developers out there in the enterprise landscape, who hold on to their code like kids to their favorite stuffed animal before bedtime**.
+
+Luckily for us, someone created a library called **JOOQ**:
+
+> jOOQ generates Java code from your database and lets you build type safe SQL queries through its fluent API
+
+<span class="image left small"><img alt="jOOQ" src="/img/SpringIO2017/jooq.png"></span>
+
+There are many use cases where jOOQ can prove useful, but the example that Michael used described the enterprise environment at his current employer:
+He was working for a **big utility company** in Germany, who developed applications on power and gas usage consisting of **lots of time-series data**.
+Almost all of the data was stored in SQL databases with a **big layer of PL/SQL on top**.
+They also developed some desktop GIS applications using **Oracle Spatial** to visualize the data.
+
+So the question they asked themselves at a certain moment:
+
+> Should we approach all of our data using plain PL/SQL? Should we use an ORM tool like Hibernate? Or can we use something in between?
+
+Of course, as with any good question, the answer is:
+
+> It depends. **There is no silver bullet**
+
+There are many options in the Java space to approach this problem:
+
+- Plain JDBC
+- Using a `JDBCTemplate`
+- JPA with JPQL and / or Criteria API
+
+Michael's team solved most of their problems using Hibernate, and while that comes with several advantages, it doesn't mean you have to use it for everything.
+One of the improvements they made was using **Hibernate combined with jOOQ**:
+
+- Use Hibernate for the regular database queries and *day-to-day manipulations of your database*
+- Use jOOQ for your complex queries and database migrations
+
+To briefly summarize what jOOQ can do for you:
+
+- jOOQ is **SQL-centric** which means jOOQ infers information from the actual database, not from the ORM model
+- It exposes a **typesafe meta-model** generated from your SQL
+- The Query Builder framework uses a **very legible DSL** (in a much more concise way than the `Criteria API`)
+- It generates a Java-based schema (using Maven or Gradle)
+- It can reverse-engineer an existing DB and generate SQL
+- Spring provides **integration with Spring Boot** through the `spring-boot-starter-jooq` dependency
+
+An ideal scenario to use jOOQ would be to:
+
+- Run your **database migration** with Flyway or Liquibase first
+- Run the **code generator** to generate the Java DSL context (this happens in the Maven `generate-sources` lifecycle phase)
+- Use the DSL context to **write your typesafe queries**, for example:
+
+{% highlight java %}
+
+BookRecord book = create.selectFrom(BOOK).where(BOOK.ID.eq(1)).fetchOne();
+
+{% endhighlight %}
+
+For more information about jOOQ, you can check out their website [here](https://www.jooq.org/){:target="_blank"}.
```

Comment:
Rule number one for SEO purposes. Don't hide links behind "here" or "click here" alike things.

Better to create a single link, e.g.: "Read more about jOOQ and what you functionality it offers."